lib/shared/s3-packages.js

```diff
@@ -0,0 +1,245 @@
+      return _.keys(_.pickBy(versions, (occurrences) => {
+        return occurrences >= NUMBER_OF_PACKAGES;
+      }));
+    })
+    .catch({
+      code: 'ENOTFOUND'
+    }, {
+      code: 'ETIMEDOUT'
+    }, () => {
+      return [];
+    });
+});
+
+/**
+ * @summary Check if a version satisfies a semver range
+ * @function
+ * @private
+ *
+ * @description
+ * This function is a wrapper around `semver.satisfies`
+ * to make it work fine with pre-release versions.
+ *
+ * @param {String} version - semver version
+ * @param {String} range - semver range
+ * @returns {Boolean} whether the version satisfies the range
+ *
+ * @example
+ * if (semverSatisfies('1.0.0', '>=1.0.0')) {
+ *   console.log('The version satisfies the range');
+ * }
+ */
+const semverSatisfies = (version, range) => {
+
+  // The `semver` module refuses to apply ranges to prerelease versions
+  // As a workaround, we drop the prerelease tags, if any, apply the range
+  // on that, and keep using the prerelease tag from then on.
+  // See https://github.com/npm/node-semver#prerelease-tags
+  const strippedVersion = `${semver.major(version)}.${semver.minor(version)}.${semver.patch(version)}`;
+
+  return semver.satisfies(strippedVersion, range);
+};
+
+/**
+ * @summary Get the latest available version for a given release type
+ * @function
+ * @public
+ *
+ * @param {String} releaseType - release type
+ * @param {Object} [options] - options
+ * @param {String} [options.range] - semver range
+ * @param {Boolean} [options.useUnstableChannel=false] - use unstable channel
+ * @fulfil {(String|undefined)} - latest version
+ * @returns {Promise}
+ *
+ * @example
+ * s3Packages.getLatestVersion(release.RELEASE_TYPE.PRODUCTION, {
+ *   range: '>=2.0.0',
+ *   useUnstableChannel: true
+ * }).then((latestVersion) => {
+ *   console.log(`The latest version is: ${latestVersion}`);
+ * });
+ */
+exports.getLatestVersion = (releaseType, options = {}) => {
+
+  // For manual testing purposes
+  const ETCHER_FAKE_S3_LATEST_VERSION = process.env.ETCHER_FAKE_S3_LATEST_VERSION;
+  if (!_.isNil(ETCHER_FAKE_S3_LATEST_VERSION)) {
+    if (release.getReleaseType(ETCHER_FAKE_S3_LATEST_VERSION) === releaseType) {
+      return Bluebird.resolve(ETCHER_FAKE_S3_LATEST_VERSION);
+    }
+
+    return Bluebird.resolve();
+  }
+
+  const bucketUrl = exports.getBucketUrlFromReleaseType(releaseType);
+  if (_.isNil(bucketUrl)) {
+    return Bluebird.reject(new Error(`No bucket URL found for release type: ${releaseType}`));
+  }
+
+  /* eslint-disable lodash/prefer-lodash-method */
+
+  return exports.getRemoteVersions(bucketUrl).filter((version) => {
+
+  /* eslint-enable lodash/prefer-lodash-method */
+
+    if (_.some([
+
+      // This check allows us to ignore snapshot builds in production
+      // buckets, and viceversa, which could have been uploaded by mistake.
+      release.getReleaseType(version) !== releaseType,
+
+      !release.isStableRelease(version) && !options.useUnstableChannel
```

Comment:
I have the feeling that lives outside the responsibilities of a semver compare function.